lib/rules/no-underscore-dangle.js

```diff
@@ -126,16 +131,55 @@ module.exports = {
         }
 
         /**
-         * Check if function has a underscore at the end
+         * Check if function parameter has a dangling underscore.
+         * @param {ASTNode} node function node to evaluate
+         * @returns {void}
+         * @private
+         */
+        function checkForDanglingUnderscoreInFunctionParameters(node) {
+            if (!allowFunctionParams) {
+                node.params.forEach(param => {
+                    let identifier = "";
+                    const { type } = param;
+
+                    if (type === "RestElement" && param.argument !== "undefined") {
```

Comment:
`param.argument !== "undefined"` condition is unnecessary. If `param.type` is `"RestElement"`, then `param.argument` cannot be `"undefined"` (nor `undefined`). It must be an `ASTNode` object.

The only thing that should be checked then is whether `param.argument` is an `"Identifier"` type node.

Maybe we could also simplify the code for all three branches with something like this:

```js
const { type } = param;
let nodeToCheck;

if (type === "RestElement") {
    nodeToCheck = param.argument;
} else if (type === "AssignmentPattern") {
    nodeToCheck = param.left;
} else {
    nodeToCheck = param;
}

if (nodeToCheck.type === "Identifier") {
    const identifier = nodeToCheck.name;

    if (hasDanglingUnderscore(identifier) && !isAllowed(identifier)) {
        // ...
    }
}
```

Reply:
I think this method looks better. Thanks to you, I learn a lot. 👍